src/Portable.Xaml/Portable.Xaml/ValueSerializerContext.cs

```diff
@@ -55,7 +55,7 @@ internal class ValueSerializerContext : IValueSerializerContext, IXamlSchemaCont
 		NamespaceResolver namespace_resolver;
 		PrefixLookup prefix_lookup;
 		XamlSchemaContext sctx;
-		IAmbientProvider ambient_provider;
+		Func<IAmbientProvider> _ambientProviderProvider;
```

Comment:
I don't think this needs to be/should be a `Func<>`, we can resolve which provider to use during the context creation.  With this, it'll actually create a new instance of the ambient provider each time it is called, which isn't very performant.

Reply:
It's not exactly that easy. I agree that initially seems to be an overkill, but it's deliberate.

This design, together with immutability of ObjectStatesAmbientProvider, guarantees ability to refer to these objects in deferred-loading situations.

As an example (and obviously reason of the change) Avalonia's DataTemplate is initiated far long after parsing XAML is finished, and only *then* is the AmbientProvider accessed. If we passed a provider that tracked changes of the parser, the object_states stack would be empty when accessed after parsing DataTemplate's container is done.

If you have an idea how to mitigate this, I'm open to suggestions. I also think it's most often completely unnecessary, but couldn't find a better solution.

Reply:
I *guess* it could be cached for each object parsed, but I'm not very proficient (read: have no idea) about the interactions, relations and dependencies between the multiple States of the Writer.

Reply:
@cwensley what do you think?